src/Simulation/qdk_sim_rs/src/python.rs

```diff
@@ -0,0 +1,243 @@
+            .map_err(|e| PyErr::new::<PyRuntimeError, String>(e))?;
+        Ok(PyState { data })
+    }
+}
+
+#[pyclass(name = "Instrument")]
+#[derive(Debug)]
+pub struct PyInstrument {
+    data: Instrument,
+}
+
+#[pyproto]
+impl PyObjectProtocol for PyInstrument {
+    fn __repr__(&self) -> String {
+        format!("<Instrument {:?}>", self.data)
+    }
+}
+
+#[pymethods]
+impl PyInstrument {
+    pub fn sample(&self, idx_qubits: Vec<usize>, state: &PyState) -> (usize, PyState) {
+        let (result, data) = self.data.sample(idx_qubits.as_slice(), &state.data);
+        (result, PyState { data })
+    }
+
+    #[staticmethod]
+    #[args(pr_readout_error = "0.0")]
+    pub fn new_z_measurement(pr_readout_error: f64) -> PyInstrument {
+        PyInstrument {
+            data: Instrument::ZMeasurement { pr_readout_error },
+        }
+    }
+
+    pub fn as_json(&self) -> String {
+        self.data.as_json()
+    }
+}
+
+#[pyclass(name = "NoiseModel")]
+#[derive(Debug)]
+pub struct PyNoiseModel {
+    data: NoiseModel,
+}
+
+#[pyproto]
+impl PyObjectProtocol for PyNoiseModel {
+    fn __repr__(&self) -> String {
+        format!("<NoiseModel {:?}>", self.data)
+    }
+}
+
+#[pymethods]
+impl PyNoiseModel {
+    #[staticmethod]
+    pub fn ideal() -> PyNoiseModel {
+        PyNoiseModel {
+            data: NoiseModel::ideal(),
+        }
+    }
+
+    #[staticmethod]
+    pub fn ideal_stabilizer() -> PyNoiseModel {
+        PyNoiseModel {
+            data: NoiseModel::ideal_stabilizer(),
+        }
+    }
+
+    #[staticmethod]
+    pub fn get_by_name(name: &str) -> PyResult<PyNoiseModel> {
+        Ok(PyNoiseModel {
+            data: NoiseModel::get_by_name(name)
+                .map_err(|e| PyErr::new::<PyRuntimeError, String>(e))?,
+        })
+    }
+
+    pub fn as_json(&self) -> String {
+        self.data.as_json()
+    }
+}
+
+// See https://stackoverflow.com/q/67412827/267841 for why the following works
+// to expose Pauli. If we have more enums like this, we could likewise expose
+// them by using the macro from that SO question.
+//
+// We got through some extra hoops to make sure everything user-facing appears
+// as a value of the Pauli enum on the Python side.
+impl IntoPy<PyObject> for Pauli {
+    fn into_py(self, py: Python) -> PyObject {
+        // Import the root module containing this native extension, and find
+        // the enum definition there that we can call.
+        let root = PyModule::import(py, "qdk_sim").unwrap();
+        let py_enum = root.get("Pauli").unwrap();
+        let args = ((self as u8).into_py(py),);
+        py_enum.call1(args).unwrap().into_py(py)
+    }
+}
+
+impl FromPyObject<'_> for Pauli {
+    fn extract(ob: &'_ PyAny) -> PyResult<Self> {
+        // We want to support either a primitive type that can extract to u8,
+        // or a value of type qdk_sim.Pauli from the root module.
+        Python::with_gil(|py| {
+            let root = PyModule::import(py, "qdk_sim").unwrap();
```

Comment:
Now that the crate's package name has changed to `qdk_sim_experimental` will this logic break? Or does this Python module name come from the library name instead of the package name?

Reply:
Good catch, this will need to change. Thanks!